dpar/src/models/tensorflow/model.rs

```diff
@@ -0,0 +1,669 @@
+    /// the graph parameters.
+    fn load_graph_<S>(
+        config_protobuf: &[u8],
+        model_protobuf: &[u8],
+        system: T,
+        vectorizer: InputVectorizer,
+        op_names: &LayerOps<S>,
+    ) -> Result<Self>
+    where
+        S: AsRef<str>,
+    {
+        let opts = ImportGraphDefOptions::new();
+        let mut graph = Graph::new();
+        graph.import_graph_def(model_protobuf, &opts)?;
+
+        let mut session_opts = SessionOptions::new();
+        session_opts.set_config(config_protobuf)?;
+        let session = Session::new(&session_opts, &graph)?;
+
+        let layer_ops = op_names.to_graph_ops(&graph)?;
+
+        let init_op = graph.operation_by_name_required(opnames::INIT)?;
+        let restore_op = graph.operation_by_name_required(opnames::RESTORE)?;
+        let save_op = graph.operation_by_name_required(opnames::SAVE)?;
+        let save_file_path_op = graph.operation_by_name_required(opnames::SAVE_FILE_PATH)?;
+
+        let is_training_op = graph.operation_by_name_required(opnames::IS_TRAINING)?;
+        let lr_op = graph.operation_by_name_required(opnames::LR)?;
+
+        let accuracy_op = graph.operation_by_name_required(opnames::ACCURACY)?;
+        let logits_op = graph.operation_by_name_required(opnames::LOGITS)?;
+        let loss_op = graph.operation_by_name_required(opnames::LOSS)?;
+        let targets_op = graph.operation_by_name_required(opnames::TARGETS)?;
+
+        let train_op = graph.operation_by_name_required(opnames::TRAIN)?;
+
+        Ok(TensorflowModel {
+            system,
+            session,
+            vectorizer,
+            layer_ops,
+            init_op,
+            restore_op,
+            save_op,
+            save_file_path_op,
+            is_training_op,
+            lr_op,
+            accuracy_op,
+            logits_op,
+            loss_op,
+            targets_op,
+            train_op,
+        })
+    }
+
+    /// Predict the best transitions for a batch of parser states.
+    ///
+    /// Both the parser states and the feature representations of the parser
+    /// states should be provided. Returns the best (possible) transition for
+    /// each parser state.
+    pub fn predict(&mut self, states: &[&ParserState], input_tensors: &LayerTensors) -> Vec<T::T> {
+        let logits = self.logits(input_tensors);
+
+        let n_labels = logits.dims()[1] as usize;
+
+        states
+            .iter()
+            .enumerate()
+            .map(|(idx, state)| {
+                let offset = idx * n_labels;
+                self.logits_best_transition(state, &logits[offset..offset + n_labels])
+            }).collect()
+    }
+
+    /// Return the best transition for a parser state.
+    ///
+    /// This method finds the best transition (largest logit) that is possible given the
+    /// current parser state.
+    fn logits_best_transition(&self, state: &ParserState, logits: &[f32]) -> T::T {
```

Comment:
these two T's are not the same and somewhat confusing. Maybe T::Transition?

Reply:
Unfortunately, the associated type is simply called T. I'll make a separate commit that fixes that.